pythonwhat/local.py

```diff
@@ -66,3 +53,46 @@ def run_code(process, code):
         raw_output = ""
         error = str(e)
     return raw_output, error
+
+
+def run_single_process(pec, code, pid=None):
+    process = StubProcess(init_code=pec, pid=pid)
+    raw_stu_output, error = run_code(process, code)
+    return process, raw_stu_output, error
+
+
+def run_exercise(pec, sol_code, stu_code, pid=None, sol_wd=None, stu_wd=None):
+    with ChDir(stu_wd or os.getcwd()):
+        stu_process, raw_stu_output, error = run_single_process(pec, stu_code, pid=pid)
```

Comment:
Just to be sure: context managers hit `__exit__` even when the body raises an error, right?

Reply:
Yes